[PyTorch] Debug-gate static_assert in KernelFunction::makeFromUnboxedFunctor

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #51368 [PyTorch] Use real if constexpr behind macro in hot template
* **#51367 [PyTorch] Debug-gate static_assert in KernelFunction::makeFromUnboxedFunctor**
* #51319 [PyTorch] Remove reference_cast in make_boxed_from_unboxed_functor
* #51463 [PyTorch] Fix return value of IValue::to for Tensor/String
* #51315 [PyTorch][easy] Fix TODOs in CppFunction constructors
* #51313 [PyTorch] Reduce template expansion in call_functor_with_args_from_stack

Templight said that this assertion was taking about 5% of build time for RegisterCPU.cpp (a hopefully-representative example I picked to shorten my iteration cycle).

I've debug-gated it on the grounds that 1) we at least try to build
everything in debug mode and 2) optimized builds presumably take
longer in general, so we can more afford to pay the build time cost in
debug builds.

The win is not entirely clear; please see the test plan for details.

Differential Revision: [D26153793](https://our.internmc.facebook.com/intern/diff/D26153793/)

swolchok committed:
[PyTorch] Debug-gate static_assert in KernelFunction::makeFromUnboxedFunctor

Templight said that this assertion was taking about 5% of build time for RegisterCPU.cpp (a hopefully-representative example I picked to shorten my iteration cycle).

I've debug-gated it on the grounds that 1) we at least try to build
everything in debug mode and 2) optimized builds presumably take
longer in general, so we can more afford to pay the build time cost in
debug builds.

The win is not entirely clear; please see the test plan for details.

Differential Revision: [D26153793](https://our.internmc.facebook.com/intern/diff/D26153793/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Debug-gate static_assert in KernelFunction::makeFromUnboxedFunctor"


Templight said that this assertion was taking about 5% of build time for RegisterCPU.cpp (a hopefully-representative example I picked to shorten my iteration cycle).

I've debug-gated it on the grounds that 1) we at least try to build
everything in debug mode and 2) optimized builds presumably take
longer in general, so we can more afford to pay the build time cost in
debug builds.

The win is not entirely clear; please see the test plan for details.

Differential Revision: [D26153793](https://our.internmc.facebook.com/intern/diff/D26153793/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Debug-gate static_assert in KernelFunction::makeFromUnboxedFunctor"


Templight said that this assertion was taking about 5% of build time for RegisterCPU.cpp (a hopefully-representative example I picked to shorten my iteration cycle).

I've debug-gated it on the grounds that 1) we at least try to build
everything in debug mode and 2) optimized builds presumably take
longer in general, so we can more afford to pay the build time cost in
debug builds.

The win is not entirely clear; please see the test plan for details.

Differential Revision: [D26153793](https://our.internmc.facebook.com/intern/diff/D26153793/)

[ghstack-poisoned]

bhosmer reviewed (approved):
This looks trivially ok - I can't imagine the build errors being related, but pinging @smessmer for thoughts [edit: ah nm it's from a prior diff]

swolchok committed:
Update on "[PyTorch] Debug-gate static_assert in KernelFunction::makeFromUnboxedFunctor"


Templight said that this assertion was taking about 5% of build time for RegisterCPU.cpp (a hopefully-representative example I picked to shorten my iteration cycle).

I've debug-gated it on the grounds that 1) we at least try to build
everything in debug mode and 2) optimized builds presumably take
longer in general, so we can more afford to pay the build time cost in
debug builds.

The win is not entirely clear; please see the test plan for details.

Differential Revision: [D26153793](https://our.internmc.facebook.com/intern/diff/D26153793/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Debug-gate static_assert in KernelFunction::makeFromUnboxedFunctor"


Templight said that this assertion was taking about 5% of build time for RegisterCPU.cpp (a hopefully-representative example I picked to shorten my iteration cycle).

I've debug-gated it on the grounds that 1) we at least try to build
everything in debug mode and 2) optimized builds presumably take
longer in general, so we can more afford to pay the build time cost in
debug builds.

The win is not entirely clear; please see the test plan for details.

Differential Revision: [D26153793](https://our.internmc.facebook.com/intern/diff/D26153793/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Debug-gate static_assert in KernelFunction::makeFromUnboxedFunctor"


Templight said that this assertion was taking about 5% of build time for RegisterCPU.cpp (a hopefully-representative example I picked to shorten my iteration cycle).

I've debug-gated it on the grounds that 1) we at least try to build
everything in debug mode and 2) optimized builds presumably take
longer in general, so we can more afford to pay the build time cost in
debug builds.

The win is not entirely clear; please see the test plan for details.

Differential Revision: [D26153793](https://our.internmc.facebook.com/intern/diff/D26153793/)

[ghstack-poisoned]

swolchok committed:
fix op_registration_test build on "[PyTorch] Debug-gate static_assert in KernelFunction::makeFromUnboxedFunctor"


Templight said that this assertion was taking about 5% of build time for RegisterCPU.cpp (a hopefully-representative example I picked to shorten my iteration cycle).

I've debug-gated it on the grounds that 1) we at least try to build
everything in debug mode and 2) optimized builds presumably take
longer in general, so we can more afford to pay the build time cost in
debug builds.

The win is not entirely clear; please see the test plan for details.

Differential Revision: [D26153793](https://our.internmc.facebook.com/intern/diff/D26153793/)

[ghstack-poisoned]
